diffrax/_step_size_controller/jump_step_wrapper.py

```diff
@@ -0,0 +1,424 @@
+            jump_ts = upcast_or_raise(
+                self.jump_ts,
+                jnp.zeros((), tdtype),
+                "`JumpStepWrapper.jump_ts`",
+                "time (the result type of `t0`, `t1`, `dt0`, `SaveAt(ts=...)` etc.)",
+            )
+
+        if self.rejected_step_buffer_len is None:
+            rejected_buffer = None
+            i_reject = jnp.asarray(0)
+        else:
+            rejected_buffer = jnp.zeros(
+                (self.rejected_step_buffer_len,) + jnp.shape(t1), dtype=tdtype
+            )
+            # rejected_buffer[len(rejected_buffer)] = jnp.inf (see def of _get_t)
+            i_reject = jnp.asarray(self.rejected_step_buffer_len)
+
+        # Find index of first element of step_ts/jump_ts greater than t0
+        i_step = _find_index(t0, step_ts)
+        i_jump = _find_index(t0, jump_ts)
+        # Clip t1 to the next element of step_ts or jump_ts
+        t1, _ = _clip_ts(t0, t1, i_step, step_ts, False)
+        t1, jump_next_step = _clip_ts(t0, t1, i_jump, jump_ts, True)
+
+        state = _JumpStepState(
+            jump_next_step,
+            dt_proposal,
+            i_step,
+            i_jump,
+            i_reject,
+            rejected_buffer,
+            step_ts,
+            jump_ts,
+            inner_state,
+        )
+
+        return t1, state
+
+    def adapt_step_size(
+        self,
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: Y,
+        y1_candidate: Y,
+        args: Args,
+        y_error: Optional[Y],
+        error_order: RealScalarLike,
+        controller_state: _JumpStepState[_ControllerState],
+    ) -> tuple[
+        BoolScalarLike,
+        RealScalarLike,
+        RealScalarLike,
+        BoolScalarLike,
+        _JumpStepState[_ControllerState],
+        RESULTS,
+    ]:
+        (
+            made_jump,
+            prev_dt,
+            i_step,
+            i_jump,
+            i_reject,
+            rejected_buffer,
+            step_ts,
+            jump_ts,
+            inner_state,
+        ) = controller_state.get()
+
+        # Let the controller do its thing
+        (
+            keep_step,
+            next_t0,
+            next_t1,
+            _,
+            inner_state,
+            result,
+        ) = self.controller.adapt_step_size(
+            t0, t1, y0, y1_candidate, args, y_error, error_order, inner_state
+        )
+
+        # This is just a logging utility for testing purposes
+        if self.callback_on_reject is not None:
+            jax.debug.callback(self.callback_on_reject, keep_step, t1)
+
+        # Check whether we stepped over an element of step_ts/jump_ts/rejected_buffer
+        # This is all still bookkeeping for the PREVIOUS STEP.
+        if step_ts is not None:
+            # If we stepped to `t1 == step_ts[i_step]` and kept the step, then we
+            # increment i_step and move on to the next t in step_ts.
+            step_inc_cond = keep_step & (t1 == _get_t(i_step, step_ts))
```

Comment:
I don't think I'm comfortable with this `==` between floating point numbers.

More generally speaking I think there could be cases in which the times being passed here do not perfectly align with the times that the adaptive step size controller suggested on the previous step (e.g. because of further wrapping of the step size controller), so I think this kind of logic is wrong anyway. I think you need something more like the `jump_ts` branch below, where you just want to snap `i_step` to the correct value. (Nothing that the correct value should be determinable statically, we only have state here for efficiency purposes.)

Reply:
I implemented a linear search as you suggest and now use it to determine `i_step` and `i_jump`.

However, I think that for the rejected step business using linear search is incorrect, because `t1` should never be greater than `rejected_t = _get_t(i_reject, rejected_buffer)`. So depending on what you think is more appropriate we can either use `==` or  `jnp.isclose(t1, rejected_t, atol=1e-12)`. I also added the code below, but I can remove it if you don't think it is necessary. I could also add a parameter that only activates this callback when we are doing tests. Let me know.

``` python
i_reject = eqx.error_if(
    i_reject,
    t1 > rejected_t,
    "Jumped over a rejected time. Please report this as a bug.",
)
```